File: go/benchmark.go
package main

import (
    "bytes"
    "fmt"
    "log"
    "os"
    "regexp"
    "time"
)

func measure(data string, pattern string) {
    start := time.Now()

    r, err := regexp.Compile(pattern)
    if err != nil {
        log.Fatal(err)
    }

    matches := r.FindAllString(data, -1)
    count := len(matches)

    elapsed := time.Since(start)

    fmt.Printf("%f - %v\n", float64(elapsed) / float64(time.Millisecond), count)
}

func main() {
    if len(os.Args) != 2 {
        fmt.Println("Usage: benchmark <filename>")
        os.Exit(1)
    }

    filerc, err := os.Open(os.Args[1])
    if err != nil {
        log.Fatal(err)
    }
    defer filerc.Close()

    buf := new(bytes.Buffer)
    buf.ReadFrom(filerc)
    data := buf.String()

    measure(data, "Twain");
    measure(data, "(?i)Twain");
    measure(data, "[a-z]shing");
    measure(data, "Huck[a-zA-Z]+|Saw[a-zA-Z]+");
    measure(data, "\\b\\w+nn\\b");
    measure(data, "[a-q][^u-z]{13}x");
    measure(data, "Tom|Sawyer|Huckleberry|Finn");
    measure(data, "(?i)Tom|Sawyer|Huckleberry|Finn");
    measure(data, ".{0,2}(Tom|Sawyer|Huckleberry|Finn)");
    measure(data, ".{2,4}(Tom|Sawyer|Huckleberry|Finn)");
    measure(data, "Tom.{10,25}river|river.{10,25}Tom");
    measure(data, "[a-zA-Z]+ing");
    measure(data, "\\s[a-zA-Z]{0,12}ing\\s");
    measure(data, "([A-Za-z]awyer|[A-Za-z]inn)\\s");
    measure(data, "[\"'][^\"']{0,30}[?!\\.][\\\"']");
    measure(data, "\\u221E|\\u2713");
    measure(data, "\\p{Sm}");
}

File: csharp_compiled/Benchmark.cs
using System;
using System.IO;
using System.Text.RegularExpressions;
using System.Diagnostics;

class Benchmark
{
    static void Main(string[] args)
    {
        if (args.Length != 1)
        {
            Console.WriteLine("Usage: benchmark <filename>");
            Environment.Exit(1);
        }

        StreamReader reader = new System.IO.StreamReader(args[0]);
        string data = reader.ReadToEnd();

        Benchmark.Measure(data, @"Twain");
        Benchmark.Measure(data, @"(?i)Twain");
        Benchmark.Measure(data, @"[a-z]shing");
        Benchmark.Measure(data, @"Huck[a-zA-Z]+|Saw[a-zA-Z]+");
        Benchmark.Measure(data, @"\b\w+nn\b");
        Benchmark.Measure(data, @"[a-q][^u-z]{13}x");
        Benchmark.Measure(data, @"Tom|Sawyer|Huckleberry|Finn");
        Benchmark.Measure(data, @"(?i)Tom|Sawyer|Huckleberry|Finn");
        Benchmark.Measure(data, @".{0,2}(Tom|Sawyer|Huckleberry|Finn)");
        Benchmark.Measure(data, @".{2,4}(Tom|Sawyer|Huckleberry|Finn)");
        Benchmark.Measure(data, @"Tom.{10,25}river|river.{10,25}Tom");
        Benchmark.Measure(data, @"[a-zA-Z]+ing");
        Benchmark.Measure(data, @"\s[a-zA-Z]{0,12}ing\s");
        Benchmark.Measure(data, @"([A-Za-z]awyer|[A-Za-z]inn)\s");
        Benchmark.Measure(data, @"[""'][^""']{0,30}[?!\.][""']");
    }

    static void Measure(string data, string pattern)
    {
        Stopwatch stopwatch = Stopwatch.StartNew();

        MatchCollection matches = Regex.Matches(data, pattern);
        int count = matches.Count;

        stopwatch.Stop();

        Console.WriteLine(stopwatch.Elapsed.TotalMilliseconds.ToString("G", System.Globalization.CultureInfo.InvariantCulture) + " - " + count);
    }
}
